### DISCUSSION TITLE
Can I use custom winUI 3 Controls in wpf application?

### DISCUSSION THREAD

**Discussion by samsunlu55atakum** [microsoft/microsoft-ui-xaml#8808] [Q&A]
Can I use custom winUI 3 Controls in wpf application? and also do I have to use microsoft.UI.xaml package to use winUI 3 controls in WPF?

**Comment by castorix** [comment]
With the last version(1.4), you can use _Microsoft.UI.Xaml.Hosting.DesktopWindowXamlSource_
I just did a quick test in Winforms/C#/.Net 7 with a Grid and a TextBlock and Initialize from _Microsoft.UI.Win32Interop.GetWindowIdFromWindow_ (same thing with WPF) :

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/22345506/712737e3-6213-4984-8fd5-20fdd6f744a4)

**Comment by DarranRowe** [comment]
For WinUI 3, you would have to use the Windows App SDK package. The [Windows App SDK is documented](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/#windows-app-sdk-features) to include WinUI 3. For hosting, you treat the Xaml Island content as a HWND which you can get through the DesktopWindowXamlSource's SiteBridge property. It uses Microsoft.UI.WindowId to represent the HWND and you need to use the GetWindowFromWindowId and GetWindowIdFromWindow to convert between them.

Right now is also not the easiest time to attempt this. First, the documentation is still incomplete. Most of the new documentation is still stubs. There is also the fact that the current implementation doesn't have any C# wrappers, this means you have to work with the raw HWND to put it into the WPF tree. Finally, if you want to use resources from XamlControlsResources, then you need to set up some extra things which you will have to implement yourself.
